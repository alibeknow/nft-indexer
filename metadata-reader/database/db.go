package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
