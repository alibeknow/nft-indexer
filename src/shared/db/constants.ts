export const MONGO_TOKENS_COLLECTION_NAME = 'tokens';
export const MONGO_CONTRACTS_COLLECTION_NAME = 'contracts';
export const MONGO_METADATA_COLLECTION_NAME = 'metadata';
export const MONGO_EVENTS_COLLECTION_NAME = 'events';
export const MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME = 'events_reader_state_log';
export const MONGO_BLOCKS_QUEUE_COLLECTION_NAME = 'blocks_queue';
export const MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME = 'lazy_mint_blocks_queue';
export const MONGO_SERVER_DUPLICATE_KEY_ERROR_CODE = 11000;
