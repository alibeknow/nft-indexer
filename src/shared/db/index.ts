export * from './constants';
export * from './db.class';
export * from './db.deprecated';
export * from './db.config';
export * from './db.module';
export * from './db.provider';
export * from './mongo-sort';

