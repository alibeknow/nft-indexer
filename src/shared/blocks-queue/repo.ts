import { Inject, Injectable } from '@nestjs/common';
import { Collection, InsertOneResult, UpdateResult, WithId } from 'mongodb';
// TODO Replace deprecated DB provider to DBProvider
import { DB } from '@shared/db';
import { MongoSort } from '../db';

export type BlockQueue = {
  from: number;
  to: number;
  toHash?: string;
  createdAt: Date;
  isDeleted?: boolean;
};

interface IDB {
  blocksQueue(): Collection<BlockQueue>;
}

export type RangeQueueAttributes = {
  from: number;
  to: number;
  toHash?: string;
  createdAt?: Date;
  isDeleted?: boolean;
};

@Injectable()
export class BlocksQueueRepo {
  constructor(
    // TODO Replace deprecated DB provider to DBProvider
    @Inject(DB) private readonly db: IDB,
  ) {}

  public async softDelete(from: number, to: number): Promise<UpdateResult> {
    return this.db.blocksQueue().updateOne({ from, to }, { $set: { isDeleted: true } });
  }

  public async save(rangeQueueAttributes: RangeQueueAttributes): Promise<InsertOneResult> {
    const { from, to, createdAt, toHash } = rangeQueueAttributes;

    return this.db.blocksQueue().insertOne({
      from,
      to,
      toHash,
      isDeleted: false,
      createdAt: createdAt ? createdAt : new Date(),
    });
  }

  public async getCreatedAfter(createdAt: Date, limit: number): Promise<WithId<BlockQueue>[]> {
    const cursor = await this.db.blocksQueue()
      .find({
        createdAt: {
          $gt: createdAt,
        },
        isDeleted: false,
      })
      .limit(limit)
      .sort({ createdAt: MongoSort.ASC });

    return cursor.toArray();
  }

  public async getLatest(to: number): Promise<WithId<BlockQueue> | null> {
    return this.db.blocksQueue()
      .findOne({
        isDeleted: false,
        to: {
          $lt: to,
        },
      }, {
        sort: { to: MongoSort.DESC },
      });
  }
}
