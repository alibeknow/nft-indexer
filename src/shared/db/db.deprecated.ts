import { Injectable } from '@nestjs/common';
import { DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';

@Injectable()
export class DB extends DBClass {
  constructor() {
    const databaseConfig: IDatabaseConfig = getDatabaseConfig();

    super(databaseConfig);
  }
}
