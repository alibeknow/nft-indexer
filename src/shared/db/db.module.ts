import { Module, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
// TODO Replace deprecated DB provider to DBProvider
import { DB } from '@shared/db';
import { logger } from '@shared/logger';

@Module({
  // TODO Replace deprecated DB provider to DBProvider
  providers: [ DB ],
  exports: [ DB ],
  imports: [],
})
export class DBModule implements OnModuleDestroy, OnModuleInit {
  constructor(
    private readonly db: DB,
  ) {}

  async onModuleInit(): Promise<void> {
    await this.db.open();
    await this.db.eventsIndexes();
    await this.db.eventsReaderStateLogIndexes();
    await this.db.createTokensIndexes();
    await this.db.createBlockQueueIndexes();
    await this.db.createLazyMintBlockQueueIndexes();
    logger.info('DB connection is opened');
  }

  async onModuleDestroy(): Promise<void> {
    await this.db.close(true);
    logger.info('DB connection is closed');
  }
}
