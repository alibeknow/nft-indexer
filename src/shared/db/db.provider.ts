import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { IBaseConfig } from '@shared/baseconfig';
import { DBClass, IDatabaseConfig } from '@shared/db';

@Injectable()
export class DBProvider<T extends IBaseConfig> extends DBClass {

  constructor(private readonly configService: ConfigService<T>) {
    super(configService.get<IDatabaseConfig>('database') as IDatabaseConfig);
  }
}
