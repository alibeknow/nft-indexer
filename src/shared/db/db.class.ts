import { LazyMintBlockQueue } from '../lazy-mint-blocks-queue';
import { Collection, MongoClient } from 'mongodb';
import { BlockQueue } from '../blocks-queue';
import {
  MONGO_BLOCKS_QUEUE_COLLECTION_NAME,
  MONGO_CONTRACTS_COLLECTION_NAME,
  MONGO_EVENTS_COLLECTION_NAME,
  MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME,
  MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME,
  MONGO_METADATA_COLLECTION_NAME,
  MONGO_TOKENS_COLLECTION_NAME,
} from './constants';
import { Token } from '../tokens';
import { Metadata } from '../metadata';
import { Event, EventsReaderStateLog } from '../events';
import { MongoSort } from './mongo-sort';
import { Contract } from '../contracts';
import { IDatabaseConfig, getMongoDBUri } from './db.config';

export class DBClass {
  private client: MongoClient;
  private directConnection: boolean;
  private sslEnable: boolean;
  private uri: string;
  private dbName: string;

  constructor(databaseConfig: IDatabaseConfig) {
    this.directConnection = databaseConfig.directConection;
    this.sslEnable = databaseConfig.enableSsl;
    this.dbName = databaseConfig.db;
    this.uri = getMongoDBUri(databaseConfig);

    let options = {};

    if(this.directConnection) {
      options = {
        ...options,
        directConnection: true,
      };
    }

    if(this.sslEnable) {
      options = {
        ...options,
        ssl: true,
        sslValidate: false,
        sslCA: `${__dirname  }/../../../cert/rds-combined-ca-bundle.pem`,
      };
    }

    this.client = new MongoClient(this.uri, options);
  }

  public async open(): Promise<MongoClient> {
    return this.client.connect();
  }

  public async close(force: boolean): Promise<void> {
    return this.client.close(force);
  }

  public tokens(): Collection<Token> {
    return this.client.db(this.dbName).collection<Token>(MONGO_TOKENS_COLLECTION_NAME);
  }

  public async createTokensIndexes(): Promise<string> {
    return this.tokens().createIndex({ createdAt: MongoSort.ASC });
  }

  public async createBlockQueueIndexes(): Promise<string> {
    return this.blocksQueue().createIndex({ to: MongoSort.DESC });
  }

  public async createLazyMintBlockQueueIndexes(): Promise<string> {
    return this.lazyMintBlocksQueue().createIndex({ to: MongoSort.DESC });
  }

  public metadata(): Collection<Metadata> {
    return this.client.db(this.dbName).collection<Metadata>(MONGO_METADATA_COLLECTION_NAME);
  }

  public events(): Collection<Event> {
    return this.client.db(this.dbName).collection<Event>(MONGO_EVENTS_COLLECTION_NAME);
  }

  public async eventsIndexes(): Promise<string> {
    return this.events().createIndex({ blockNumber: MongoSort.ASC, logIndex: MongoSort.ASC });
  }

  public eventsReaderStateLog(): Collection<EventsReaderStateLog> {
    return this.client.db(this.dbName).collection<EventsReaderStateLog>(MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME);
  }

  public async eventsReaderStateLogIndexes(): Promise<string> {
    return this.eventsReaderStateLog().createIndex({ startAt: MongoSort.DESC });
  }

  public blocksQueue(): Collection<BlockQueue> {
    return this.client.db(this.dbName).collection<BlockQueue>(MONGO_BLOCKS_QUEUE_COLLECTION_NAME);
  }

  public lazyMintBlocksQueue(): Collection<LazyMintBlockQueue> {
    return this.client.db(this.dbName).collection<LazyMintBlockQueue>(MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME);
  }

  public contracts(): Collection<Contract> {
    return this.client.db(this.dbName).collection<Contract>(MONGO_CONTRACTS_COLLECTION_NAME);
  }
}
