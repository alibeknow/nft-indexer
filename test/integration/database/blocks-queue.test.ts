import { WithId } from 'mongodb';
import { BlockQueue, BlocksQueueRepo } from '@shared/blocks-queue';
import { DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';

jest.setTimeout(100000);

describe('Blocks queue repo', () => {
  let db: DBClass;
  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `db-blocks-queue-repo-test-${dbConfig.db}` });
    await db.open();
  });

  describe('[save] method', () => {
    it('should save a document to MongoDB in range queue collection', async () => {
      const from = 14316460;
      const to = 14316495;
      const toHash = '0x55836c255338b545522ceac0966e4a04da387bd84e2a4c830c28feb07693de2a';
      const rangeQueueRepo = new BlocksQueueRepo(db);
      const saveResult = await rangeQueueRepo.save({
        from,
        to,
        toHash,
        createdAt: new Date('2019-12-15T04:55:00'),
      });

      const docs = await db.blocksQueue().find({ _id: saveResult.insertedId }).toArray();
      // To get rid of non-permanent field _id
      // eslint-disable-next-line
      expect(docs.map(({ _id, ...rest }) => ({ ...rest }))).toMatchSnapshot();
    });
  });

  describe('[getCreatedAfter] method', () => {
    it('should retrieve range queue documents created after a specific date', async () => {
      const rangeQueueRepo = new BlocksQueueRepo(db);

      const ranges = [
        {
          from: 14316461,
          to: 14316465,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-15T04:55:00'),
        },
        {
          from: 14316466,
          to: 14316470,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-16T02:22:00'),
        },
        {
          from: 14316471,
          to: 14316475,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-17T03:24:00'),
        },
        {
          from: 14316476,
          to: 14316480,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-17T03:25:00'),
        },
        {
          from: 14316481,
          to: 14316485,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-17T03:26:00'),
        },
      ];

      await Promise.all(ranges.map(
        (range) => rangeQueueRepo.save(range),
      ));

      const result = await Promise.all([
        rangeQueueRepo.getCreatedAfter(new Date('2019-12-17T00:00:00'), 10),
        rangeQueueRepo.getCreatedAfter(new Date('2019-12-17T04:00:00'), 10),
        rangeQueueRepo.getCreatedAfter(new Date('2018-12-17T04:00:00'), 10),
        rangeQueueRepo.getCreatedAfter(new Date('2019-12-16T02:22:00'), 10),
        rangeQueueRepo.getCreatedAfter(new Date('2019-12-17T03:26:00'), 10),
      ]);

      // To get rid of non-permanent field _id
      // eslint-disable-next-line
      const prepareResults = (items: WithId<BlockQueue>[]) => items.map(({ _id, ...rest }) => ({ ...rest }));

      expect(prepareResults(result[0])).toHaveLength(3);
      expect(prepareResults(result[0])).toMatchSnapshot();
      expect(prepareResults(result[1])).toHaveLength(0);
      expect(prepareResults(result[1])).toMatchSnapshot();
      expect(prepareResults(result[2])).toHaveLength(5);
      expect(prepareResults(result[2])).toMatchSnapshot();
      expect(prepareResults(result[3])).toHaveLength(3);
      expect(prepareResults(result[3])).toMatchSnapshot();
      expect(prepareResults(result[4])).toHaveLength(0);
      expect(prepareResults(result[4])).toMatchSnapshot();
    });
  });

  describe('[getLatest] method', () => {
    it('should retrieve the latest range queue document', async () => {
      const rangeQueueRepo = new BlocksQueueRepo(db);

      const ranges = [
        {
          from: 14316461,
          to: 14316465,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-15T04:55:00'),
        },
        {
          from: 14316466,
          to: 14316470,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-18T02:22:00'),
        },
        {
          from: 14316471,
          to: 14316471,
          toHash: 'fake-hash',
          createdAt: new Date('2019-12-17T03:24:00'),
        },
      ];

      await Promise.all(ranges.map(
        (range) => rangeQueueRepo.save(range),
      ));

      const result = await rangeQueueRepo.getLatest(14316471);

      // To get rid of non-permanent field _id
      /* eslint-disable */
      // @ts-ignore
      const { _id, ...rest } = result;
      /* eslint-enable */

      expect(result?.createdAt).toEqual(new Date('2019-12-18T02:22:00'));
      expect(rest).toMatchSnapshot();
    });
  });

  describe('[softDelete] method', () => {
    it('should update block range queue with isDeleted = true', async () => {
      const rangeQueueRepo = new BlocksQueueRepo(db);
      const from = 14316461;
      const to = 14316465;

      const saveResult = await rangeQueueRepo.save({
        from,
        to,
        toHash: 'fake-hash',
        createdAt: new Date('2019-12-15T04:55:00'),
      });
      await rangeQueueRepo.softDelete(from, to);
      const doc = await db.blocksQueue().findOne({ _id: saveResult.insertedId });

      // To get rid of non-permanent field _id
      /* eslint-disable */
      // @ts-ignore
      const { _id, ...rest } = doc;
      /* eslint-enable */

      expect(rest).toMatchSnapshot();
    });
  });

  afterEach(async () => {
    await db.blocksQueue().deleteMany({});
  });

  afterAll(async () => {
    await db.close(true);
  });
});
