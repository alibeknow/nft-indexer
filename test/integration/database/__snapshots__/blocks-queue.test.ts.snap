// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocks queue repo [getCreatedAfter] method should retrieve range queue documents created after a specific date 1`] = `
[
  {
    "createdAt": 2019-12-17T03:24:00.000Z,
    "from": 14316471,
    "isDeleted": false,
    "to": 14316475,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:25:00.000Z,
    "from": 14316476,
    "isDeleted": false,
    "to": 14316480,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:26:00.000Z,
    "from": 14316481,
    "isDeleted": false,
    "to": 14316485,
    "toHash": "fake-hash",
  },
]
`;

exports[`Blocks queue repo [getCreatedAfter] method should retrieve range queue documents created after a specific date 2`] = `[]`;

exports[`Blocks queue repo [getCreatedAfter] method should retrieve range queue documents created after a specific date 3`] = `
[
  {
    "createdAt": 2019-12-15T04:55:00.000Z,
    "from": 14316461,
    "isDeleted": false,
    "to": 14316465,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-16T02:22:00.000Z,
    "from": 14316466,
    "isDeleted": false,
    "to": 14316470,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:24:00.000Z,
    "from": 14316471,
    "isDeleted": false,
    "to": 14316475,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:25:00.000Z,
    "from": 14316476,
    "isDeleted": false,
    "to": 14316480,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:26:00.000Z,
    "from": 14316481,
    "isDeleted": false,
    "to": 14316485,
    "toHash": "fake-hash",
  },
]
`;

exports[`Blocks queue repo [getCreatedAfter] method should retrieve range queue documents created after a specific date 4`] = `
[
  {
    "createdAt": 2019-12-17T03:24:00.000Z,
    "from": 14316471,
    "isDeleted": false,
    "to": 14316475,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:25:00.000Z,
    "from": 14316476,
    "isDeleted": false,
    "to": 14316480,
    "toHash": "fake-hash",
  },
  {
    "createdAt": 2019-12-17T03:26:00.000Z,
    "from": 14316481,
    "isDeleted": false,
    "to": 14316485,
    "toHash": "fake-hash",
  },
]
`;

exports[`Blocks queue repo [getCreatedAfter] method should retrieve range queue documents created after a specific date 5`] = `[]`;

exports[`Blocks queue repo [getLatest] method should retrieve the latest range queue document 1`] = `
{
  "createdAt": 2019-12-18T02:22:00.000Z,
  "from": 14316466,
  "isDeleted": false,
  "to": 14316470,
  "toHash": "fake-hash",
}
`;

exports[`Blocks queue repo [save] method should save a document to MongoDB in range queue collection 1`] = `
[
  {
    "createdAt": 2019-12-15T04:55:00.000Z,
    "from": 14316460,
    "isDeleted": false,
    "to": 14316495,
    "toHash": "0x55836c255338b545522ceac0966e4a04da387bd84e2a4c830c28feb07693de2a",
  },
]
`;

exports[`Blocks queue repo [softDelete] method should update block range queue with isDeleted = true 1`] = `
{
  "createdAt": 2019-12-15T04:55:00.000Z,
  "from": 14316461,
  "isDeleted": true,
  "to": 14316465,
  "toHash": "fake-hash",
}
`;
